libraries/lib-wave-track/Sequence.cpp

```diff
@@ -1114,6 +1114,23 @@ bool Sequence::Read(samplePtr buffer, sampleFormat format,
    return true;
 }
 
+LibSampleTrack::ClipBlockSequence
+Sequence::GrabFloats(sampleCount start, sampleCount length) const
+{
+   const auto limitedLength = std::min(length, GetNumSamples() - start);
+   std::vector<LibSampleTrack::BlockPtr> blocks;
+   const auto sequenceOffset = start - GetBlockStart(start);
+   auto cursor = start;
+   while (cursor < start + length)
+   {
+      const auto b = FindBlock(cursor);
+      const SeqBlock& block = mBlock[b];
+      blocks.push_back(block.sb->Get());
+      cursor = block.start + block.sb->GetSampleCount();
+   }
+   return { std::move(blocks), sequenceOffset.as_size_t(), length };
```

Comment:
This holds shared pointers to blocks, so there is no dangling pointer problem.   But this might construct a cache that becomes out of date with changing sequence contents.

Maybe it won’t matter in actual uses but comments should explain the limitation.

Or maybe a complete redesign of caching with have Sequence maintain more state and keep consistency.

Reply:
I understand better how it works now.

Every time you call `Get` on the decorator, it goes back to the Sequence and finds the blocks that overlap the range.  The old cache is not yet destroyed when a new one is made.  Any blocks with a non-expired weak pointer to a vector of samples use that.

If the underlying Sequence has been changed between calls to `Get`, there is no invalidation problem.  The range of the request is just mapped to the new blocks.